Cabal/Distribution/PackageDescription/Parsec.hs

```diff
@@ -0,0 +1,330 @@
+
+    -- ** Supplementary build information
+    -- readHookedBuildInfo,
+    --parseHookedBuildInfo,
+    ) where
+
+#if __GLASGOW_HASKELL__ < 710
+import Data.Monoid         (Monoid(..))
+import Control.Applicative (Applicative(..), (<$>))
+#endif
+
+import Data.Maybe    (listToMaybe, isJust, maybeToList)
+import Distribution.License
+import Control.Applicative (Alternative((<|>)))
+import Data.Char     (isSpace)
+import Data.Foldable (traverse_)
+import Data.Maybe    (listToMaybe, isJust)
+import Data.List     (nub, unfoldr, partition, (\\))
+import Control.Monad (liftM, foldM, when, unless, ap)
+import Control.Arrow    (first)
+import System.Directory (doesFileExist)
+import Data.Typeable
+import Data.Data
+import Text.PrettyPrint
+import Data.Map (Map)
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BS.Char8
+import qualified Data.Map as Map
+
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State
+
+import Distribution.Text
+         ( Text(disp, parse, parsec), display, simpleParsec )
+import Distribution.Version (VersionRange)
+import Language.Haskell.Extension
+
+import Distribution.PrettyUtils
+import Distribution.PackageDescription
+import Distribution.PackageDescription.Utils
+         ( cabalBug, userBug )
+import Distribution.Package
+         ( PackageIdentifier(..), Dependency(..), packageName, packageVersion )
+import Distribution.ModuleName ( ModuleName )
+import Distribution.Version
+        ( Version(Version), orLaterVersion
+        , LowerBound(..), asVersionIntervals )
+import Distribution.Verbosity (Verbosity)
+import Distribution.Compiler  (CompilerFlavor(..))
+import Distribution.PackageDescription.Configuration (freeVars)
+import Distribution.Simple.Utils
+         ( die, dieWithLocation, warn, intercalate, lowercase, cabalVersion
+         , withFileContents, withUTF8FileContents
+         , writeFileAtomic, writeUTF8File )
+
+import Distribution.Parsec.Parser
+import Distribution.Parsec.Types.FieldDescr
+import Distribution.Parsec.Types.Common (Position(..), showPos, showPWarning)
+import Distribution.Parsec.Types.ParseResult
+import Distribution.Parsec.FieldDescr
+import Distribution.Parsec.Utils
+import Distribution.Parsec.Types.Common
+
+import qualified Text.Parsec as Parsec
+
+-- ---------------------------------------------------------------
+-- Parsing
+
+-- | Given a parser and a filename, return the parse of the file,
+-- after checking if the file exists.
+readAndParseFile
+    :: (BS.ByteString -> ParseResult a)
+    -> Verbosity
+    -> FilePath
+    -> IO a
+readAndParseFile parser verbosity fpath = do
+    exists <- doesFileExist fpath
+    unless exists
+        (die $ "Error Parsing: file \"" ++ fpath ++ "\" doesn't exist. Cannot continue.")
+    bs <- BS.readFile fpath
+    case parser bs of
+        PR ws (Left errs) -> do
+            mapM_ (warn verbosity . showPWarning fpath) $ reverse ws
+            -- | TODO: die properly
```

Comment:
I plan to make parser accumulate all errors it could, so we'll need to print them all and die in different way.